lib/backend/pgbk/atomicwrite.go

```diff
@@ -0,0 +1,161 @@
+
+		for _, ca := range condacts {
+			switch ca.Condition.Kind {
+			case backend.KindWhatever:
+				// no comparison to assert
+			case backend.KindExists:
+				condBatch.Queue(
+					"SELECT EXISTS (SELECT FROM kv WHERE key = $1 AND (expires IS NULL OR expires >= now()))",
+					nonNil(ca.Key),
+				).QueryRow(func(row pgx.Row) error {
+					var cond bool
+					if err := row.Scan(&cond); err != nil {
+						return trace.Wrap(err)
+					}
+					conditionFailed = conditionFailed || !cond
+					return nil
+				})
+			case backend.KindNotExists:
+				condBatch.Queue(
+					"SELECT NOT EXISTS (SELECT FROM kv WHERE key = $1 AND (expires IS NULL OR expires >= now()))",
+					nonNil(ca.Key),
+				).QueryRow(func(row pgx.Row) error {
+					var cond bool
+					if err := row.Scan(&cond); err != nil {
+						return trace.Wrap(err)
+					}
+					conditionFailed = conditionFailed || !cond
+					return nil
+				})
+			case backend.KindRevision:
+				expectedRevision, ok := revisionFromString(ca.Condition.Revision)
+				if !ok {
+					return struct{}{}, trace.Wrap(backend.ErrConditionFailed)
+				}
+				condBatch.Queue(
+					"SELECT EXISTS (SELECT FROM kv WHERE key = $1 AND revision = $2 AND (expires IS NULL OR expires >= now()))",
+					nonNil(ca.Key), expectedRevision,
+				).QueryRow(func(row pgx.Row) error {
+					var cond bool
+					if err := row.Scan(&cond); err != nil {
+						return trace.Wrap(err)
+					}
+					conditionFailed = conditionFailed || !cond
+					return nil
+				})
+			default:
+				return struct{}{}, trace.BadParameter("unexpected condition kind %v in conditional action against key %q", ca.Condition.Kind, ca.Key)
+			}
+
+			switch ca.Action.Kind {
+			case backend.KindNop:
+				// no action to be taken
+			case backend.KindPut:
+				includesPut = true
+				actBatch.Queue(
+					"INSERT INTO kv (key, value, expires, revision) VALUES ($1, $2, $3, $4)"+
+						" ON CONFLICT (key) DO UPDATE SET"+
+						" value = excluded.value, expires = excluded.expires, revision = excluded.revision",
+					nonNil(ca.Key), nonNil(ca.Action.Item.Value), zeronull.Timestamptz(ca.Action.Item.Expires.UTC()), newRevision,
+				)
+			case backend.KindDelete:
+				actBatch.Queue(
+					"DELETE FROM kv WHERE kv.key = $1 AND (kv.expires IS NULL OR kv.expires > now())", nonNil(ca.Key),
+				)
+			default:
+				return struct{}{}, trace.BadParameter("unexpected action kind %v in conditional action against key %q", ca.Action.Kind, ca.Key)
+			}
+		}
+
+		tx, err := b.pool.Begin(ctx)
+		if err != nil {
+			return struct{}{}, trace.Wrap(err)
+		}
+
+		// Rollback is designed to be safe to defer and will have no effect if we end up
+		// committing the transaction before returning.
+		defer tx.Rollback(ctx)
+
+		if condBatch.Len() != 0 {
+			if err := tx.SendBatch(ctx, &condBatch).Close(); err != nil {
+				return struct{}{}, trace.Wrap(err)
+			}
+		}
+
+		if conditionFailed {
+			return struct{}{}, trace.Wrap(backend.ErrConditionFailed)
+		}
```

Comment:
(unless we start returning `ErrConditionFailed` from `condBatch`'s callbacks)

```suggestion
		if condBatch.Len() != 0 {
			if err := tx.SendBatch(ctx, &condBatch).Close(); err != nil {
				return struct{}{}, trace.Wrap(err)
			}
			if conditionFailed {
				return struct{}{}, trace.Wrap(backend.ErrConditionFailed)
			}
		}
```